### PR TITLE
Refactor convolution_backward's cudnn cases

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71492 Refactor convolution_backward Miopen cases
* **#71491 Refactor convolution_backward's cudnn cases**
* #71490 Refactor convolution_backward's CudaDepthwise3d case
* #71489 Refactor convolution_backward's CudaDepthwise2d case

Changed the Cudnn and CudnnTranspose cases to only make the input
contiguous when it is needed for the grad_weight computation.

Reading the implementation of cudnn_convolution_transpose_backward and
cudnn_convolution_backward give me confidence that `input` isn't used
for the grad_weight computation. However, the memory format logic is so
convoluted that I'm not 100% sure this is correct. All the tests pass though
and on request I can directly pass `backend_memory_format` to
{cudnn_convolution_backward, cudnn_convolution_transpose_backward}.

Test Plan:
- pytest test/test_nn.py -v -k "conv"

Differential Revision: [D33664694](https://our.internmc.facebook.com/intern/diff/D33664694)

### CODE REVIEW

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33664694).

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33664694).

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -1638,7 +1638,9 @@ std::tuple<Tensor, Tensor, Tensor> convolution_backward(
       std::array<bool, 2> input_weight_output_mask = {output_mask[0], output_mask[1]};
       std::tie(backend_grad_input, backend_grad_weight) = cudnn_convolution_backward_stub(
           input.device().type(),
-          input.contiguous(backend_memory_format), grad_output, weight, params.padding, params.stride,
+          // Only make input contiguous when it is necessary for the backwards computation
+          output_mask[1] ? input.contiguous(backend_memory_format) : input,
+          grad_output, weight, params.padding, params.stride,
```
Comment (jbschlosser) [inline aten/src/ATen/native/Convolution.cpp:1643]: My understanding is that this covers the case where:
* `weight` is `channels_last`
* `input` is **not** `channels_last`
* `grad_output` is **not** `channels_last`
* `grad_weight` is being computed, which involves input & grad_output

We want channels_last to be used, but `cudnn_convolution_backward_weight` bases its decision only on input / grad_output:
https://github.com/pytorch/pytorch/blob/f45e217c0158c247772f6491a18662cc972c29cd/aten/src/ATen/native/cudnn/ConvShared.cpp#L387

Explicitly setting `input.contiguous(backend_memory_format)` before the call to `cudnn_convolution_backward_weight` will trigger the above logic to return channels_last.

Interestingly, we *don't* need analogous logic for `cudnn_convolution_backward_input` to do the right thing because of:
https://github.com/pytorch/pytorch/blob/f45e217c0158c247772f6491a18662cc972c29cd/aten/src/ATen/native/cudnn/ConvShared.cpp#L430

and

https://github.com/pytorch/pytorch/blob/f45e217c0158c247772f6491a18662cc972c29cd/aten/src/ATen/native/cudnn/ConvShared.cpp#L470

ensuring channels_last is correctly used in this case:
* `weight` is **not** `channels_last`
* `input` is `channels_last`
* `grad_output` is **not** `channels_last` (will be made channels_last due to `input.suggest_memory_format()` calls above)
* `grad_input` is being computed, which involves weight & grad_output

TL;DR: Changes lgtm :)

**Review by jbschlosser — APPROVED** [review]
